### PR TITLE
eng: add support for snapshot tests

### PR DESCRIPTION
**Pull request by connor4312**
This adds Jest-like support for snapshot testing.
Developers can do something like:

```js
await assertSnapshot(myComplexObject)
```

The first time this is run, the snapshot expectation file is written
to a `__snapshots__` directory beside the test file. Subsequent runs
will compare the object to the snapshot, and fail if it doesn't match.

You can see an example of this in the test for snapshots themselves!

After a successful run, any unused snapshots are cleaned up. On a failed
run, a gitignored `.actual` snapshot file is created beside the
snapshot for easy processing and inspection.

Shortly I will do some integration with the selfhost test extension to
allow developers to easily update snapshots from the vscode UI.

For #189680

cc @ulugbekna @hediet

### CODE REVIEW

File: test/unit/browser/renderer.html [hunk 1]
```diff
@@ -116,15 +116,22 @@
 			runner.on('pending', test => window.mocha_report('pending', serializeRunnable(test)));
 		};
 
+
+		async function loadModules(modules) {
```
Comment (connor4312) [inline test/unit/browser/renderer.html:120]: Changing how we load files was necessary for Mocha to populate the `currentTest.file`, which we need to know where to place snapshots

**Review by ulugbekna — DISMISSED** [review]
This is super cool, excited to see how this looks in UI! 

A couple of questions: 

1. Anything I can help with? 
2. Do you think we could also support inline snapshots?
3. Can a snapshot be a multi-line string? Is it deindented? 
4. Are we going to support updating snapshots on file/suite/test/assertion levels? 

(nice touch with snapception :))

**Comment by connor4312** [comment]
Tried tossing some fixes at components I saw in the stacktraces but I doubt they're the culprit. Every test I've added the disposable tracker to has had leaks. @hediet what do you want to do here?

**Comment by connor4312** [comment]
I started going up the tests happening before the diff editor tests. Inline chat has lots of undisposed disposables. I spent an hour or so just getting fixes for those in, but that's not really sustainable to do everywhere (also wildly out of scope for this PR.) I also don't think we should block doing improvements in test infrastructure (this and #173056) behind it.

In this PR, I will comment out the problematic line. I think fixing it and re-inserting it is good, but I don't want to say we can't have snapshot testing until some indeterminable leak gets fixed...

**Comment by hediet** [comment]
For the diff editor tests, I do this: https://github.com/microsoft/vscode/tree/cb94a70e1f43b4115e86907165c80d7870f10e18/src/vs/editor/test/node/diffing

That way, running the tests already updates the snapshots and you can use the diff editor from the SCM view to review the changes (for the diff algorithm snapshots I have a custom editor that renders the snapshotted diff visually).
To accept the changes, you just delete the *.invalid files.

This works very well in practice.

**Comment by hediet** [comment]
> In this PR, I will comment out the problematic line. I think fixing it and re-inserting it is good, but I don't want to say we can't have snapshot testing until some indeterminable leak gets fixed...

I kind of agree, but with this change, you basically made the indeterminable leak my problem now ([which I already investigated for a couple of hours](https://github.com/microsoft/vscode/pull/173056#issuecomment-1470238702))... Maybe we should come up with a plan for the team to tackle this together.

**Review by M4ximumPizza — COMMENTED** [review]
Looks super cool. I am excited to see how this turns out in the end.

**Comment by connor4312** [comment]
Yea, I will make an adoption item to use the leak detector in tests people own

**Review by joyceerhl — APPROVED** [review]

**Review by alexr00 — APPROVED** [review]

**Pull request merged**